cmd/otel-allocator/allocation/least_weighted.go

```diff
@@ -0,0 +1,195 @@
+
+	"github.com/go-logr/logr"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	_            Allocator = &leastWeightedAllocator{}
+	strategyName           = "least-weighted"
+)
+
+/*
+	Load balancer will serve on an HTTP server exposing /jobs/<job_id>/targets
+	The targets are allocated using the least connection method
+	Load balancer will need information about the collectors in order to set the URLs
+	Keep a Map of what each collector currently holds and update it based on new scrape target updates
+*/
+
+// leastWeightedAllocator makes decisions to distribute work among
+// a number of OpenTelemetry collectors based on the number of targets.
+// Users need to call SetTargets when they have new targets in their
+// clusters and call SetCollectors when the collectors have changed.
+type leastWeightedAllocator struct {
+	// m protects collectors and targetItems for concurrent use.
+	m sync.RWMutex
+	// collectors is a map from a Collector's name to a Collector instance
+	collectors map[string]*Collector
+	// targetItems is a map from a target item's hash to the target items allocated state
+	targetItems map[string]*TargetItem
+
+	log logr.Logger
+}
+
+// TargetItems returns a shallow copy of the targetItems map.
+func (allocator *leastWeightedAllocator) TargetItems() map[string]*TargetItem {
+	allocator.m.RLock()
+	defer allocator.m.RUnlock()
+	targetItemsCopy := make(map[string]*TargetItem)
+	for k, v := range allocator.targetItems {
+		targetItemsCopy[k] = v
+	}
+	return targetItemsCopy
+}
+
+// Collectors returns a shallow copy of the collectors map.
+func (allocator *leastWeightedAllocator) Collectors() map[string]*Collector {
+	allocator.m.RLock()
+	defer allocator.m.RUnlock()
+	collectorsCopy := make(map[string]*Collector)
+	for k, v := range allocator.collectors {
+		collectorsCopy[k] = v
+	}
+	return collectorsCopy
+}
+
+// findNextCollector finds the next collector with fewer number of targets.
+// This method is called from within SetTargets and SetCollectors, whose caller
+// acquires the needed lock. This method assumes there are is at least 1 collector set
+func (allocator *leastWeightedAllocator) findNextCollector() *Collector {
+	var col *Collector
+	for _, v := range allocator.collectors {
+		// If the initial collector is empty, set the initial collector to the first element of map
+		if col == nil {
+			col = v
+		} else if v.NumTargets < col.NumTargets {
+			col = v
+		}
+	}
+	return col
+}
+
+// addTargetToTargetItems assigns a target to the next available collector and adds it to the allocator's targetItems
+// This method is called from within SetTargets and SetCollectors, which acquire the needed lock.
+// This is only called after the collectors are cleared or when a new target has been found in the tempTargetMap
+func (allocator *leastWeightedAllocator) addTargetToTargetItems(target *TargetItem) {
+	chosenCollector := allocator.findNextCollector()
```

Comment:
This can be nil if `len(allocator.collectors) == 0`.  Validating and short-circuiting in this case is probably necessary to avoid nil pointer errors.

Reply:
Kristina brought up a similar concern for the `findNextCollector()`, this issue existed before and was covered by the fact that no collectors isn't currently a valid state. For right now, i think we need to introduce the same short circuit to verify the invariant for `SetTargets` instead of introducing error handling here. After this is merged, I'll open up an issue to solve this in a better way.